share/spack/setup-env.sh

```diff
@@ -60,21 +60,43 @@ _spack_shell_wrapper() {
         emulate -L sh
     fi
 
-    # accumulate flags meant for the main spack command
-    # the loop condition is unreadable, but it means:
-    #     while $1 is set (while there are arguments)
-    #       and $1 starts with '-' (and the arguments are flags)
+    # accumulate flags meant for spack itself preceding the command
     _sp_flags=""
-    while [ ! -z ${1+x} ] && [ "${1#-}" != "${1}" ]; do
-        _sp_flags="$_sp_flags $1"
-        shift
+    help_or_version_was_used=0
+    while [ -n "$1" ]; do
+        case "$1" in
+            --color|--config|--sorted-profile|--lines|--env-dir|--env)
+                # long flags taking required values
+                _sp_flags="$_sp_flags $1 $2"
+                shift 2
+                ;;
+            -c*|-e*|-D*)
```

Comment:
I think this will fail to parse the (invalid) command `spack -Cd /path/to/config ...`. Will it still report a reasonable error message in that case?

Reply:
let me check, it will interpret /path... as the command.

Also what's missing is `-C` actually :D missed that one.

Reply:
Ugh. Of course `spac -klde [env]` _is_ allowed :( so it needs nested iteration to match each character

Reply:
or is it not? I'm just thinking of `tar -xzf [file]` which is equivalent to `tar -x -z -f [file]` iirc

Reply:
Yeah `spack -de foo find` is allowed.

Reply:
ok, that's supported now

Reply:
if there is a better way to do shell support that doesn't require argument parsing for 3 different shells that'd be great

Reply:
@becker33 it does what you expect now:

```$ spack -Cd /path/to/config hello
==> Error: config scope is not a directory: 'd'
$ spack -dC /path/to/config hello
==> [2021-03-17-22:10:32.023095] ConfigError: config scope is not a directory: '/path/to/config'
==> [2021-03-17-22:10:32.023676] Error: config scope is not a directory: '/path/to/config'
...
spack.config.ConfigError: config scope is not a directory: '/path/to/config'
```